Add a electrum syncing example

rajarshimaitra opened the pull request:
This PR adds an electrum syncing example.. Fixes #72 

The electrum sync is done in two steps.
 - `wallet_txid_scan` returns a candidate `KeychainScan` with TxGraph data.
 - The candidate is compared with tracker's `SparseChain`, and list of new tx data to fetch is derived from the `ChangeSet`.
 - New transactions are fetched in batch and added into the `KeychainScan` update.
 - Apply the final `KeychainScan` on tracker and DB.

rajarshimaitra commented:
Restructured the commit to handle #72 requirements.

 - Make Chaingraph consistent, by adding few assertions at the `apply_update()` API. 
 - Make the electrum sync in two steps. Get all the txids, find new ones, fetch tx data, apply update..

rajarshimaitra commented:
Me and @evanlinjin tend to disagree on the approach of the consistency assertion.. While to me it seems the best place to apply it is the update API, Evan thinks there could be a better approach.. Opening this up for discussion and reviews..

rajarshimaitra commented:
Rebased on current master, and updated the example syncing with recent invariants around `ChainGraph`. Similar to RPC syncing, the first stage returns a `SparseChain`. The nex `to_fetch_tx` id set is derived using `determine_changeset` between the returned sparsechain and existing sparaechain within the tracker..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,150 @@
+
+/// We will detect reorg if it has depth less than this
+const REORG_DETECTION_DEPTH: u32 = 100;
+const DEFAULT_STOP_GAP: usize = 10;
```
There should be no default stop gap.

LLFourn replied:
No if there's a default you just don't stop.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,150 @@
+
+        // 1. Create checkpoint data from last_known_height - REORG_DEPTH
+        // Reorg situation will he handled at the time of applying this KeychainScan.
+        // If there's reorg deeper than the assumed depth, update process will throw error.
```
I guess this comment is stale? I don't see an error thrown.

rajarshimaitra replied:
The intention was to say that the error will occur while performing `apply_update` with the `KeychainScan`.. Thats what I meant by the `update process will throw error.`.. Maybe can be made more clear?

LLFourn replied:
Ah sorry I didn't read this. Let's design it so this error case doesn't exist :). I don't see how you're meant to recover from the error with this API.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,150 @@
+        let current_height = self.get_height()?;
+        let check_from = last_known_height
+            .map(|ht| ht.saturating_sub(REORG_DETECTION_DEPTH))
+            .unwrap_or(0);
```
If the tip at `last_known_height` has changed then this will always lead to an error in the case the checkpoint before `last_known_height` is more than 100 blocks away. Why not do it the way we did in esplora?

rajarshimaitra replied:
Yes, I was going with the assumption that `SparseChain` will always maintain consecutive sequence of blocks.. 

The rationale behind that was, in sync code we are always adding sequential blocks in checkpoints, and pruning is happening from the start of the chain.. So by that logic we should never have the last common at 100+ blocks away from `last_known_height`? As long as that's true, the other way of throwing error here is more than 100 blocks of reorg, which is unlikely..

But if my assumption is wrong then yes this should not be done in this way, and the esplora way is better. This one is clearly less elegant than the esplora way, but went in this route to reduce passing in any refs from the tracker.. The chain was the only refs required to pass in in the sync function..

LLFourn replied:
> Yes, I was going with the assumption that `SparseChain` will always maintain consecutive sequence of blocks..

This is definitely not the case. Perhaps it's confusing to call it a "chain". It's meant to be a view of a chain that is sparse not a chain that is sparse.

> But if my assumption is wrong then yes this should not be done in this way, and the esplora way is better. This one is clearly less elegant than the esplora way, but went in this route to reduce passing in any refs from the tracker.. The chain was the only refs required to pass in in the sync function..

You can just clone the checkpoints to avoid taking a ref. There might be other ways to design this process but the way it works is to limit the number of checkpoints in the sparse chain with `checkpoint_limit` so they can be quickly cloned and passed in here without taking a ref. This seems to be ideal since the your checkpoint limit will be a function of your confidence of the persistence of the underlying blockchain and the memory available on your platform -- these are not things that can change at runtime and should be fixed in the config.

rajarshimaitra replied:
Okay.. Ya that makes sense to me.. So I will revert it back to the esplora way..

evanlinjin reviewed (changes requested):
Good work so far!

rajarshimaitra commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,150 @@
+                    .collect::<Vec<(Txid, TxHeight)>>();
+
+                if history.is_empty() {
+                    unused_script_count += 1;
```
At the start of the loop for every keychain..

LLFourn commented on bdk_electrum_example/src/main.rs:
```diff
@@ -0,0 +1,169 @@
+            let changeset = tracker.determine_changeset(&scan_update)?;
+            db.append_changeset(&changeset)?;
+            tracker.apply_changeset(changeset);
+            debug!("sync completed!!")
```
Let's put the duplicated code between sync and scan into a function.

rajarshimaitra replied:
Good idea.. Was thinking exactly that..

LLFourn replied:
It seems like there is still a bit of duplicate code. Isn't it the case that everything after getting the sparse chain and applying the last active indexes will be the same?

LLFourn replied:
I think you can at least just return the `new_sparsechain` from the match arms and do the rest outside of the match statement.

rajarshimaitra commented on bdk_electrum_example/src/main.rs:
```diff
@@ -0,0 +1,169 @@
+            // New Txids that we don't know the full data of.
+            let txid_changeset = tracker
+                .chain()
+                .determine_changeset(&new_sparsechain)?
```
I don't follow.. Its a `SparseChain::ChangeSet`, that only gives the `txid` changes. So we don't have tx_additions at this point. 

But I see your point, and added another filter to not fetch redundant transactions in the sync. But this causes another complexity, as in theory we can have partial transactions in the graph, and `insert_tx` only allows full transaction data.

I have added the new logic in latest push, and marked the above as TODO.

rajarshimaitra commented:
Updated with the above suggestions.

 - No default reorg depth mechanism. Takes in a ref of existing chain, and inserts checkpoint until the last common check point is found. And then push the current tip into update. In case of reorg, all checkpoints after the last common block is added into the update structure, and while `SparseChain::apply_update` new checkpoints will replace old checkpoints. No error is thrown.
 
 - No default stop gap.
 
 - Transaction fetching is done opportunistically, for only txids that we haven't seen before. For transactions that only changed in `Index` tx data is fetched from existing `TxGraph` of the tracker. 
 
 Q: What to do when we only have `TxNode::Partial` in graph, and index changes of that transaction? 2 possible options. 
  - We can fetch such transactions again from the blockchain before applying update.
  - Have mechanism to `insert_tx()` with `TXNode::Partial`
  
- A new function added in the main to collect all the common code between `Scan` and `Sync`.

LLFourn commented:
> Q: What to do when we only have `TxNode::Partial` in graph, and index changes of that transaction? 2 possible options.
> 
>     * We can fetch such transactions again from the blockchain before applying update.
> 
>     * Have mechanism to `insert_tx()` with `TXNode::Partial`

Haven't reviewed just comment: we don't keep the index for `TxNode::Partial`.

rajarshimaitra commented:
> Haven't reviewed just comment: we don't keep the index for TxNode::Partial.

Done..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,151 @@
+            }
+
+            for (txid, index) in script_history_txid {
+                sparse_chain.insert_tx(txid, index)?;
```
Deal with this error here.

LLFourn replied:
You are throwing the error on `insert_checkpoint` so doesn't this just return with an error if there was a re-org (since `insert_checkpoint` returns an error if the hashes don't match). It seems simpler just to check that the hash at the start tip height is the same and throw an error if not.

It looks like you are trying to do your best even in the face of a re-org and still return a sparse chain. This is a cool idea. It might require some more thought though. For now I suggest just returning a `ReOrg` error if there was a re-org while you are doing the scan and we can make an issue about how to do better than this.

Inserting some extra checkpoints is ok -- we do that in esplora but in that case we have a single call to fetch the most recent ones. It might not be worth it here.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,151 @@
+                    .script_get_history(&script)?
+                    .iter()
+                    .map(|history_result| {
+                        if history_result.height > 0 {
```
```suggestion
                        if history_result.height > 0 && history_result.height <= tip_height {
```

we don't want to add as confirmed things above the tip height. Fine to add them to unconfirmed. This just means a new block has been confirmed since you started. I'm not sure what we did in esplora but we should try and do the same thing there too.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,151 @@
+
+            keychain_index_update.insert(keychain, last_active_index);
+        }
+
```
Must check with electrum that the chain at the end still contains `tip_hash`. If not return an error indicating there has been a re-org during scan.

rajarshimaitra replied:
The in sync reorg situation is handled as described above.. https://github.com/LLFourn/bdk_core_staging/pull/77#discussion_r1049572798

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,151 @@
+                    height: existing_height,
+                    hash: current_hash,
+                })
+                .expect("should not collide");
```
This is a very interesting "should". When does this panic?

LLFourn commented on bdk_electrum_example/src/main.rs:
```diff
@@ -0,0 +1,201 @@
+
+            // Inserting everything from the new_sparsechain should be okay as duplicate
+            // data would be rejected at the time of update application.
+            for (height, hash) in new_sparsechain.checkpoints() {
```
Why doesn't this just return `BlockId`s :thinking: . I swear it used to.

LLFourn commented on bdk_electrum_example/src/main.rs:
```diff
@@ -0,0 +1,201 @@
+fn main() -> anyhow::Result<()> {
+    let (args, keymap, mut tracker, mut db) = bdk_cli::init::<ElectrumCommands, _>()?;
+
+    let client = ElectrumClient::new("ssl://electrum.blockstream.info:60002")?;
```
needs to be different for different networks?

rajarshimaitra replied:
I feel stupid but I can't seem to find any docs out there for electrum with other network ports.. Any help??

LLFourn reviewed (changes requested):
Thanks @rajarshimaitra. Looks great. A few minor requests to finish this off. The main thing that's missing is to protect against re-org during sync.

rajarshimaitra commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,151 @@
+                height: tip_height,
+                hash: tip_hash,
+            })
+            .expect("Should not collide");
```
Same as above..

rajarshimaitra replied:
Hmm.. I think so ya.. Returning a Reorg error in that case??

rajarshimaitra commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,151 @@
+        scripts: BTreeMap<K, impl Iterator<Item = (u32, Script)>>,
+        stop_gap: Option<usize>,
+        local_chain: &BTreeMap<u32, BlockHash>,
+    ) -> Result<(SparseChain, BTreeMap<K, u32>)> {
```
Added custom result.. Does it make sense to use `thiserror` for this case? Would reduce boiler plates.. If yes, will do that on a separate PR.

LLFourn replied:
I don't really mind whether we use thiserror in the blockchain example modules.

rajarshimaitra commented:
> It seems like there is still a bit of duplicate code. Isn't it the case that everything after getting the sparse chain and applying the last active indexes will be the same?

Yes.. I tried to add this into a separate function but that's making the code more complicated, because many trait bounds needs to be satisfied for the `KeychainStore::append_changeset()`, which are handled in `bdk_cli`.. It still could be done, but after doing it it made the code more complex than it needs to be, and thought would be detriment for readers reading this as example, and its only removing 3 lines. So kept it as it is for now..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,236 @@
+            let mut unused_script_count = 0usize;
+            let mut script_history_txid = Vec::<(Txid, TxHeight)>::new();
+
+            loop {
```
This should be a `while let Some((index, script)) = scripts.next()`

rajarshimaitra replied:
Done..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,236 @@
+                        .iter()
+                        .map(|history_result| {
+                            if history_result.height > 0
+                                && (history_result.height as u32) < tip_height
```
`(history_result.height as u32) <= tip_height`

rajarshimaitra replied:
Done..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,236 @@
+    Client(electrum_client::Error),
+    InsertTx(InsertTxErr),
+    Update(UpdateFailure),
+    InsertCheckpoint(InsertCheckpointErr),
```
Each of these errors needs docs. It should be clear how a user is meant to recover from each of them. If I get a client error I guess I know that something went wrong connecting to the electrum client but with the others I have no clue what it means.

rajarshimaitra replied:
Removed these error variants in latest push..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,236 @@
+                        unused_script_count += 1;
+                    } else {
+                        last_active_index = index;
+                        script_history_txid.extend(history.iter());
```
why not just insert into the sparse chain right here.

rajarshimaitra replied:
Good point.. Done..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,236 @@
+            loop {
+                if let Some((index, script)) = scripts.next() {
+                    let history = self
+                        .script_get_history(&script)?
```
shouldn't we be using `batch_script_get_history` for efficiency here?

rajarshimaitra replied:
We can, but currently we are getting all the scripts using `iter_all_script_pubkeys_by_keychain()`, which calls  `descriptor_into_script_iter()` which is an iterator over all the scripts that can be derived from a descriptor. So if we call `batch_script_history`, it will work for the `Sync` call, but for `Scan` it will try to fetch for `2^32` values for wild card descriptors which can be a pretty massive call..

LLFourn replied:
Sorry I missed that comment. Look at what we do in electrum impl in bdk already. You just take a bunch of scripts up to `batch_size` which the caller can pass in and batch it in that way.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,236 @@
+                if let Err(err) = sparse_chain.insert_tx(txid, index) {
+                    match err {
+                        InsertTxErr::TxTooHigh => {
+                            /* Don't care about new transactions confirmed while syncing */
```
since you checked that no tx is higher than tip I think this should be `unreachable!(..)`.

rajarshimaitra replied:
Makes sense.. Done..

rajarshimaitra commented:
> Each of these errors needs docs. It should be clear how a user is meant to recover from each of them. If I get a client error I guess I know that something went wrong connecting to the electrum client but with the others I have no clue what it means.

On further thinking these error variants should not be there. As users don't know what to do in those cases. I think in the previous code we could have just `expect` the result because those were emitted from an empty sparsechain.

> It looks like you are trying to do your best even in the face of a re-org and still return a sparse chain. This is a cool idea. It might require some more thought though. For now I suggest just returning a ReOrg error if there was a re-org while you are doing the scan and we can make an issue about how to do better than this.

Yes that was my initial intention. As users won't know what to do in case of a in-sync reorg. And I don't think there's any recourse available for user in the current API. So ideally  `wallet_txid_scan` should only throw client error, and everything else should be handled by the logic itself. In case of reorg we should be able to still return a sparsechain..

But I went ahead with your suggestion and removed that extra logic, and now it just throws a reorg error in case reorg is detected. Added a TODO to note this point for further works.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,205 @@
+                height: tip_height,
+                hash: tip_hash,
+            }).is_err() {
+                // This means our existing chain tip has been reorged out.
```
It means the tip you just added to has changed since you last requested it at the end of the for loop above. You use "existing" to refer to the one in `local_chain` above but that's not what's happening here. Please clear up this comment.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,205 @@
+                            match err {
+                                InsertTxErr::TxTooHigh => {
+                                    /* We should not encounter this error as we ensured TxHeight <= tip_height */
+                                    unreachable!();
```
Put the comment above in the unreachable statement as a string.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,205 @@
+            {
+                return Err(ElectrumError::Reorg(height));
+            };
+        }
```
Simplify this. I think we should just check that our tip hasn't changed here. If it has return the re-org error. That's enough for now.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,205 @@
+#[derive(Debug)]
+pub enum ElectrumError {
+    Client(electrum_client::Error),
+    Reorg(u32),
```
I don't think this height thing is useful. What is the caller meant to do with it?

Some doc comments here would be good to explain how to recover from the two different error kinds. `Reorg` should just be call the function again.

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,205 @@
+
+            while let Some((index, script)) = scripts.next() {
+                let history = self
+                    .script_get_history(&script)?
```
Lets use `batch_script_get_history` here.

rajarshimaitra replied:
I didn't do this one because of the comment here https://github.com/LLFourn/bdk_core_staging/pull/77#discussion_r1050440828 and also couldn't figure how to use `stop_gap` in that case. Let me know if you intend to fetch for 2^32 scripts anyway and should we remove stop gap?? 

The other way is by calling `script_pubkeys_by_keychain` which will give us iterator over only derived scripts.. Let me know if that's the way we wanna go here..

LLFourn replied:
Sorry I responded in the linked convo.

LLFourn commented on bdk_electrum_example/src/main.rs:
```diff
@@ -0,0 +1,213 @@
+            // Fetch the new and old transactions to be added in update structure
+            for (tx, index) in fetch_transactions(&new_sparsechain, &client, &tracker)? {
+                scan_update.update.insert_tx(tx, index)?;
+            }
```
Can you try simplifying the above with the new APIs we added in https://github.com/LLFourn/bdk_core_staging/pull/92?

You can use `SpaseChain::changeset_additions` to figure out what the changeset is between your local sparse chain and the new sparse chain.. Fetch those full transactions and then call `ChainGraph::inflate_changeset` to turn your `sparse_chain::ChangeSet` into a `ChainGraph::ChangeSet`. You can then set the `KeychainChangeSet`'s `chain_graph` field with this value and just apply that changeset.

LLFourn replied:
Thanks this looks great.

LLFourn reviewed (commented):
I just took another pass at this. The list of things I've proposed to fix here is final now I think.

rajarshimaitra commented:
Thanks @LLFourn for the comments.. Updated as suggested and used the new API.. modulo https://github.com/LLFourn/bdk_core_staging/pull/77#discussion_r1054936229..  Code looks much simpler..

LLFourn commented on bdk_electrum_example/src/main.rs:
```diff
@@ -0,0 +1,189 @@
+    let sparsechain_changeset = match tracker.chain().determine_changeset(new_sparsechain) {
+        Ok(cs) => cs,
+        Err(update_failure) => {
+            return Err(anyhow!(
```
just use `Anyhow::Context::context` to add these annotations rather than matching them.

rajarshimaitra replied:
The major reason I couldn't apply `context` or just pass on the error is the error type returned here isn't `std::err:Error`. So compiler is complaining on trait bound.. That's the reason for such awkward error handling.. Same with other places where the return type isn't `std::Error`..  

Here's the full error
```
error[E0599]: the method `context` exists for enum `Result<(), (KeychainChangeSet<Keychain, _>, HashSet<bdk_core::bitcoin::Txid>)>`, but its trait bounds were not satisfied
   --> bdk_electrum_example/src/main.rs:116:57
    |
116 |             tracker.apply_changeset(keychain_changeset).context("Changeset Application to tracker failed")?;
    |                                                         ^^^^^^^ method cannot be called on `Result<(), (KeychainChangeSet<Keychain, _>, HashSet<bdk_core::bitcoin::Txid>)>` due to unsatisfied trait bounds
    |
   ::: /home/raj/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:504:1
    |
504 | pub enum Result<T, E> {
    | --------------------- doesn't satisfy `_: bdk_cli::anyhow::Context<(), (KeychainChangeSet<Keychain, _>, HashSet<bdk_core::bitcoin::Txid>)>`
    |
    = note: the following trait bounds were not satisfied:
            `(KeychainChangeSet<Keychain, _>, HashSet<bdk_core::bitcoin::Txid>): anyhow::context::ext::StdError`
            which is required by `Result<(), (KeychainChangeSet<Keychain, _>, HashSet<bdk_core::bitcoin::Txid>)>: bdk_cli::anyhow::Context<(), (KeychainChangeSet<Keychain, _>, HashSet<bdk_core::bitcoin::Txid>)>`
```
The work around for that is either this awkward handling, or create dedicated error type to return those errors. I think we should do the later, but would prefer to do them in a separate PR..

LLFourn commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,197 @@
+            BlockId { height, hash }
+        };
+
+        if sparse_chain.latest_checkpoint().expect("must exist") != tip_blockid {
```
This tip changing doesn't mean there has been a re-org. The block hash at the same height as `latest_checkpoint` changing means there has been a re-org.

rajarshimaitra replied:
Yes makes sense.. Updated with hash matching of latest checkpoint that we have..

LLFourn reviewed (commented):
Look great @rajarshimaitra. Main things:

- use batch calls to electrum
-  Intead of matching every error and then throwing a new anyhow error just use `anyhow::Context::context`
- Remove function `complete_update` in favor of just getting the `new_sparsechain` from the match arms and doing the update from there like what @evanlinjin does here: https://github.com/LLFourn/bdk_core_staging/pull/107/files#diff-6f57072cac1464050942f5bf972a9a8f76cb82a151f93f5b5d16cab1515f925aR74

rajarshimaitra commented:
@LLFourn updated with the suggestions, module https://github.com/LLFourn/bdk_core_staging/pull/77#discussion_r1056064131. 

Let me know if this looks good and you want the the commit history squashed..

rajarshimaitra commented:
Sqaushed and rebased on master..

LLFourn commented:
I just did some work to get this working:

- break at the right time in the loop for stop gap
- add signet electrum server
- print out scnning progress
- Make client configurable
- Stop returning the sparse chain from inflate_changeset because it
  makes it harder to ? the error and you can just clone it if you still
  need it on failure.

Check it out @rajarshimaitra.

This is working and probably ready to merge with final review from @evanlinjin.

evanlinjin commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,212 @@
+                                    TxHeight::Confirmed(history_result.height as u32),
+                                )
+                            } else {
+                                (history_result.tx_hash, TxHeight::Unconfirmed)
```
Smart.

rajarshimaitra commented:
ACK https://github.com/LLFourn/bdk_core_staging/commit/35319072cdfc2de982a4ccb69714c6aa1a601203

Thanks, @LLFourn for the fixes.. Code looks much better.. 

Added a nit fix.. 

Testing out the full thing..

evanlinjin commented:
@rajarshimaitra You forced-pushed away the changes...

rajarshimaitra commented:
Holy shit..

evanlinjin commented:
@rajarshimaitra Fixed.

rajarshimaitra commented:
Sorry dude.. Intended to just push a nit.. Fat fingered..

evanlinjin commented on bdk_electrum_example/src/electrum.rs:
```diff
@@ -0,0 +1,212 @@
+                    if txid_list.is_empty() {
+                        unused_script_count += 1;
+                    } else {
+                        last_active_index = index;
```
Given the API, it is not guaranteed that the scripts will be given in order by index (`impl Iterator<Item = (u32, Script)>`). Should we check if `index` is greater than `last_active_index` before changing?

evanlinjin commented on bdk_electrum_example/src/main.rs:
```diff
@@ -0,0 +1,191 @@
+}
+
+#[derive(Parser, Debug, Clone, PartialEq)]
+pub struct ScanOption {
+    /// Set batch size for each script_history call to electrum client
+    #[clap(long, default_value = "25")]
+    pub batch_size: usize,
+}
```
NIT: If we can add an optional electrum_url flag, it would be helpful imo.

evanlinjin replied:
I can't find a way to do this elegantly so I will skip this.

evanlinjin reviewed (approved):
ACK 35319072cdfc2de982a4ccb69714c6aa1a601203

Just some nits
